Add docker deploy

shibaobun opened the pull request:
Please let me know if you'd like anything to be changed in this PR!

## Features
- Dockerizes the toolkit with multiple use cases 
  - Schedule commands via cron
  - Run one-off commands with `docker exec`
  - Add `command` configuration in `docker-compose.yml`
- Adds docker instructions to readme

## To do
- Add docker hub username and password to [Github Action Secrets](https://github.com/reluce/szurubooru-toolkit/settings/secrets/actions) to enable automatic builds

reluce commented:
Two things I'd like to add:
* Can you rename the szuru-toolkit references to szurubooru-toolkit? Just so we're consistent.
* Is it possible to run the one-off commands without prefixing `poetry run` everytime? Does the `.bashrc` get read when using `docker container exec`? If so, setting an alias would probably be sufficient: `alias auto-tagger='poetry run auto-tagger'`

I'm not too familiar with GitHub actions, but I trust you that things "just work" after I've added the secrets to the repo 😉.

shibaobun commented:
Unfortunately this image just builds off the base python image (I couldn't get the alpine image working no matter how hard I tried), which doesn't have bash, only sh. I'll look into seeing what I can do with the aliases though

shibaobun commented:
Hi, I renamed everything from szuru to szurubooru, and I also added the scripts to the path so all the scripts are now accessible without having to use `poetry run` :smile:
Just another note, I replaced the username in `docker-compose.yml` with `reluce`. So if you make a Docker hub account with a different username, you'll need to update the `image:` config in `docker-compose.yml` with that username instead. Thank you!

reluce commented:
I've cloned your branch to test some things. Here's what I observed:

The `$PATH` solution isn't the best, as the scripts don't have a shebang which points to Python. After building the image, running the container with bash was no problem: `docker container exec -it <container_name> bash`. I've set an alias in `/root/.bashrc` with the content `alias auto-tagger="poetry run auto-tagger"` and it does seem to work fine. You can probably append each script as a line to the `.bashrc` in the `Dockerfile`. 

Another thing which I didn't get to work were the bind volumes (I'm using WSL2 with Ubuntu). After executing `docker-compose up -d`, I always got:
```
Starting szurubooru-toolkit ... error

ERROR: for szurubooru-toolkit  Cannot start service szurubooru-toolkit: failed to create shim: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error mounting "/home/user/repos/szuru-github/test/szurubooru-toolkit/config.toml" to rootfs at "/szurubooru-toolkit/config.toml": mount /home/user/repos/szuru-github/test/szurubooru-toolkit/config.toml:/szurubooru-toolkit/config.toml (via /proc/self/fd/6), flags: 0x5000: not a directory: unknown: Are you trying to mount a directory onto a file (or vice-versa)? Check if the specified host path exists and is the expected type

ERROR: for szurubooru-toolkit  Cannot start service szurubooru-toolkit: failed to create shim: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error mounting "/home/user/repos/szuru-github/test/szurubooru-toolkit/config.toml" to rootfs at "/szurubooru-toolkit/config.toml": mount /home/user/repos/szuru-github/test/szurubooru-toolkit/config.toml:/szurubooru-toolkit/config.toml (via /proc/self/fd/6), flags: 0x5000: not a directory: unknown: Are you trying to mount a directory onto a file (or vice-versa)? Check if the specified host path exists and is the expected type
ERROR: Encountered errors while bringing up the project.
```

After removing the `VOLUME` instructions in the `Dockerfile`, the volumes mounted correctly without errors. Are these still somewhere necessary? Also, the `/temp` directory isn't necessary as well.

Aside from that, everthing else looks good to me.

shibaobun commented:
Huh, I must've been misremembering when I was still trying to get it to work with the alpine image. Docker also has some weirdness about sourcing `.bashrc` since it doesn't open a bash shell by default, I think this post explains it a bit more: https://stackoverflow.com/a/55213158
However, I did add the aliases to the entrypoint script that gets run if the user doesn't change the `command:` configuration in `docker-compose.yml`, but if they're trying to change the startup command then they can probably just use `poetry run` instead, however when I was testing it, the python scripts were running fine even without the bash bang in front of the scripts.

As for the docker volume error, that can happen when you start the container by adding a bind mapping to `config.toml` without it already existing, which docker will then create a directory in that location automatically instead. You can fix it by running `docker-compose stop` and then `docker-compose rm -v` which will remove the volume mappings and let you recreate the mappings by doing `docker-compose up -d` again.

Finally, the `VOLUME` commands in the dockerfile don't really do much actually, they mark that those directories should be volumes and that docker internally should create volume directories for them. However, this isn't really needed in this case so I'll remove them from the dockerfile, since it can be pretty easy for people to remove the container without removing the corresponding volumes and they won't update if the upstream container updates, apparently which I just learned :smile:

reluce commented on Dockerfile:
```diff
@@ -0,0 +1,32 @@
+RUN poetry config virtualenvs.create false && \
+  poetry install --only main --ansi --no-interaction
+
+ENV PATH="${PATH}:/szurubooru-toolkit/src/scripts"
```
This isn't necessary as `poetry install` makes the scripts available under `/usr/local/bin`.

shibaobun replied:
oh nice, that would explain why it was just working lol

reluce commented on entrypoint.sh:
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env sh
+
+# add aliases
+alias auto-tagger="poetry run auto-tagger"
+alias create-tags="poetry run create-tags"
+alias delete-posts="poetry run delete-posts"
+alias import-from-booru="poetry run import-from-booru"
+alias reset-posts="poetry run reset-posts"
+alias upload-media="poetry run upload-media"
+alias tag-posts="poetry run tag-posts"
```
Can be removed as well (see above).

reluce commented on README.md:
```diff
@@ -22,6 +22,37 @@ Alternatively, you can clone the package from GitHub and set everything up with
+   in `crontab_sample`.
+1. If you would like to use deepbooru or tag files, create `misc/deepbooru`
+   and/or `misc/tags` in the same location and follow the instructions linked
+   below
```
Another bind volume is necessary for the `src_path` in the `config.toml`, otherwise, the script won't start.

Something like:

> 1. Edit the `docker-compose.yml` file and add another entry under `volumes` for your `src_path` specified in `config.toml`. It needs to be in the following format: `- <local_src_path>:<target_path>`. It's recommended to specify the target path as `/szurubooru-toolkit/upload_src`. In that case, the `src_path` in your `config.toml` should point to `/szurubooru-toolkit/upload_src`.

shibaobun replied:
ooh good point, I'll replace the temp mapping with that

reluce commented on README.md:
```diff
@@ -22,6 +22,37 @@ Alternatively, you can clone the package from GitHub and set everything up with
+1. If you would like the container to run a one-time command and then quit with
+   `docker-compose.yml`, add a `command` configuration [like
+   this](https://docs.docker.com/compose/compose-file/compose-file-v3/#command).
+   You may also need to prepend `poetry run` to the command as well.
```
I believe this line isn't needed anymore as well 😉 .

reluce commented on README.md:
```diff
@@ -22,6 +22,38 @@ Alternatively, you can clone the package from GitHub and set everything up with
+   the commands you would like to run regularly. An example command is provided
+   in `crontab_sample`.
+1. Make sure to set the `src_path` option in `config.toml` to use
+   `/szurubooru-toolkit/upload_src`.
```
I'd add that the user needs to modify the `/upload_src` in `docker-compose.yml`. This would also be the only time the user has to interact with that file.

shibaobun replied:
I also changed the temp directory in docker-compose.yml to be upload_src since it shouldn't be needed for anything, right?

reluce replied:
Yeah that's fine. I think I kept it as a reference in some places just in case, but I'll remove these soon as anything temp related is being executed in RAM.

reluce replied:
Would you be so kind and add the note for the `docker-compose.yaml` file I mentioned? I can then merge this PR 😅 .

shibaobun replied:
Hi, what note do you need? Docker should automatically create a folder at that location, so the only change needed should be setting that directory in `config.toml`, is there anything else needed?

reluce replied:
The current `docker-compose.yml` specifies on line 12  `- ./upload_src:/szurubooru-toolkit/upload_src`. The user needs to set `/szurubooru-toolkit/upload_src` in `config.toml`, yes - but he might also have to change `./upload_src` in case he has another dedicated directory for that (well, at least I do 😉 ).

So, the note should just tell the user that he might need to change the `./upload_src` in the `docker-compose.yml` file.

shibaobun replied:
Sorry for the delay, but I've added the note now, please let me know if there are any other changes you'd like!
